### PR TITLE
feat: Support multiple redirectUris for OAuth

### PR DESCRIPTION
**Pull request by sheff146**
Added new fields. Updated authenticate() method. Updated manifest validation and translation. Updated documentation.

### CODE REVIEW

File: docs/outreach-api.md [hunk 1]
```diff
@@ -32,10 +32,10 @@ access rights with the [scopes](scopes.md) defined in the add-on manifest.
 Once a user consent to that and authorize Outreach API access, [initial authentication flow](#oauth-sequence-diagram)
 will start.
 
-A request to the authorisation endpoint defined in [redirectUri](manifest.md#redirecturi) will be made with a **"code"**
-query parameter value sent from Outreach authentication server. This code is a short-lived authorization token used with
-[Outreach application](manifest.md#applicationid), and Outreach OAuth app secret so a proper Outreach API access token
-and refresh tokens could be obtained.
+A request to the authorisation endpoint defined in one of [redirectUris](manifest.md#redirecturis) will be made with a
+**"code"** query parameter value sent from Outreach authentication server. This code is a short-lived authorization
+token used with [Outreach application](manifest.md#applicationid), and Outreach OAuth app secret so a proper Outreach
```
Comment (palootcenas-outreach) [inline docs/outreach-api.md:37]:
```suggestion
token used with [Outreach application](manifest.md#client), and Outreach OAuth app secret so a proper Outreach
```

File: docs/outreach-api.md [hunk 2]
```diff
@@ -52,11 +52,11 @@ Outreach API access:
...
 
    - Addon opens popup using the Outreach [API authentication URL](https://api.outreach.io/api/v2/docs#authentication).
-     Constructing of this url relies on manifest.api section configuration: [applicationId](manifest.md#applicationid),
-     [redirectUri](manifest.md#redirecturi) and [scopes](manifest.md#scopes) values.
+     Constructing of this url relies on manifest.api section configuration: [client.id](manifest.md#client),
+     [redirectUris](manifest.md#redirecturis) and [scopes](manifest.md#scopes) values.
    - When user clicks **Authorize** button on [Outreach consent screen](assets/api-consent.png), Outreach will redirect
-     to `/authorize` endpoint on ([manifest.api.redirectUri](manifest.md#redirecturi)) address with a short living
-     authorization code passed as query param.
+     to `/authorize` endpoint on one of ([manifest.api.redirectUris](manifest.md#redirecturis)) address with a short
```
Comment (palootcenas-outreach) [inline docs/outreach-api.md:58]: What you mean by "one of"? Is there some precise rule which URL is taken? I.e. tries first URL with 2s timeout then tries the next, etc. ? Or does it call random one?

Comment (palootcenas-outreach) [reply docs/outreach-api.md:58]: I see now from documentation below.

File: docs/outreach-api.md (hunk 1 above)
Comment (sheff146) [reply docs/outreach-api.md]: Nice catch! Thanks!

File: docs/schema/2.0/manifest.schema.json [hunk 3]
```diff
@@ -69,7 +69,9 @@
               "type": "string"
             }
```
Comment (palootcenas-outreach) [inline docs/schema/2.0/manifest.schema.json:70]: I'm wondering if we are versioning the schema somehow?
I see the `$id` might be usable if we picked SemVer. 
E.g. instead of 
```
"$id": "https://github.com/getoutreach/extensibility-sdk/docs/schema/2.0/manifest.schema.json",
```
...we'd use `2.0.1`
```
"$id": "https://github.com/getoutreach/extensibility-sdk/docs/schema/2.0.1/manifest.schema.json",
```

Or did we decide not to create a new version unless a breaking change is introduced?

File: src/legacy/ManifestTranslator.ts [hunk 4]
```diff
@@ -182,7 +179,17 @@ export class ManifestTranslator {
     app.store.type = storeType;
 
     app.extensions = [];
-    app.api = firstExt.api;
+    if (firstExt.api) {
+      app.api = new ManifestApi();
+      app.api.client = new ManifestApiClient();
+      app.api.scopes = firstExt.api.scopes;
+      app.api.client.id = firstExt.api.applicationId;
+      app.api.redirectUris = [firstExt.api.redirectUri];
```
Comment (palootcenas-outreach) [inline src/legacy/ManifestTranslator.ts:187]: Shouldn't this take into account also `redirecUris` ?
I.e.
```suggestion
      app.api.redirectUris = firstExt.api.redirectUris ?? [firstExt.api.redirectUri];
```

File: src/sdk/Validator.ts [hunk 5]
```diff
@@ -26,83 +24,72 @@ export const validate = (application: Application): string[] => {
...
       issues.push('Manifest Api section needs to have applicationId value.');
     }
 
+    if (!application.api.client) {
+      issues.push('Manifest Api section needs to have client value.');
+    } else if (!application.api.client.id) {
+      issues.push('Manifest Api section needs to have client.id value.');
+    }
+
     if (!utils.urlValidation(application.api.redirectUri)) {
```
Comment (palootcenas-outreach) [inline src/sdk/Validator.ts:40]: This should validate with regard to `application.api.redirectUris`. 
E.g. now it will always fail on missing `redirectUri` even though the `redirectUris` is filled.

File: src/sdk/services/oauthService.ts [hunk 6]
```diff
@@ -24,11 +22,28 @@ class OAuthService {
     return originUrl.origin.replace(regex[1], 'accounts');
   };
 
-  private getOAuthAuthorizeUrl = (api: ManifestApi) => {
+  private getOAuthAuthorizeUrl = (api: ManifestApi, redirectUri?: string) => {
     const host = this.getOAuthHost();
-    const scopes = encodeURIComponent(api.scopes.join(' '));
-    const redirectUri = encodeURIComponent(api.redirectUri);
-    return `${host}/oauth/authorize?client_id=${api.applicationId}&redirect_uri=${redirectUri}&response_type=code&scope=${scopes}`;
+    const scopes = encodeURIComponent(api.scopes.join(','));
```
Comment (palootcenas-outreach) [inline src/sdk/services/oauthService.ts:27]: Isn't this a breaking change? Or it was never used so far?

File: src/sdk/services/oauthService.ts [hunk 7]
```diff
@@ -24,11 +22,28 @@ class OAuthService {
...
+    const selectedRedirectUri = encodeURIComponent(this.selectRedirectUri(api, redirectUri));
+    const clientId = api.client.id ?? api.applicationId;
```
Comment (palootcenas-outreach) [inline src/sdk/services/oauthService.ts:29]: Should be encoded as well.
```suggestion
    const clientId =  encodeURIComponent(api.client.id ?? api.applicationId);
```

File: src/sdk/services/oauthService.ts [hunk 8]
```diff
@@ -24,11 +22,28 @@ class OAuthService {
...
+    return `${host}/oauth/authorize?client_id=${clientId}&redirect_uri=${selectedRedirectUri}&response_type=code&scope=${scopes}`;
+  };
+
+  private selectRedirectUri = (api: ManifestApi, redirectUri?: string) => {
+    if (redirectUri === undefined) {
+      if (api.redirectUris.length) {
```
Comment (palootcenas-outreach) [inline src/sdk/services/oauthService.ts:35]: What if `api.redirectUris` is `undefined` ?

File: src/legacy/ManifestTranslator.ts (hunk 4 above)
Comment (sheff146) [reply src/legacy/ManifestTranslator.ts:187]: It is a mapping of V1 manifests to V2 manifests. There is no redirectUris in V1.

File: src/sdk/services/oauthService.ts (hunk 6 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts]: I don't know why there was a space here. I checked with accounts service, they actually use `+` to concatenate scopes, so comma was wrong. Thanks for bringing this into attention.

File: src/manifest/ManifestApi.ts [hunk 9]
```diff
@@ -34,16 +37,16 @@ export class ManifestApi {
...
   /**
    * Address of the endpoint, which will return support refresh token flow.
    *
    * @type {string}
-   * @memberof Api
-   * @deprecated Usage of token endpoint is deprecated and it will be removed in upcoming releases.
+   * @memberof ManifestApi
+   * @deprecated Usage of token endpoint is deprecated and will be removed in upcoming releases.
    */
   token: string;
```
Comment (palootcenas-outreach) [inline src/manifest/ManifestApi.ts:51]: Shouldn't we make it also options when deprecating? That applies to all of the deprecated props here.

File: src/sdk/services/oauthService.ts (hunk 8 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts:35]: We have TypeScript and runtime validation to help us ensure it's not. But I can add an additional check.

File: src/sdk/services/oauthService.ts [hunk 10]
```diff
@@ -2,14 +2,12 @@ import { ManifestApi } from '../../manifest/ManifestApi';
 import runtime from '../RuntimeContext';
 
 class OAuthService {
```
Comment (palootcenas-outreach) [inline src/sdk/services/oauthService.ts:4]: `OAuthService` is missing tests. 
Maybe we should create some, especially when we're doing big changes like this one.

File: src/sdk/services/oauthService.ts (hunk 7 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts]: Fixed

File: src/sdk/services/oauthService.ts (hunk 8 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts]: Added

File: src/legacy/ManifestTranslator.ts [hunk 11]
```diff
@@ -32,17 +32,15 @@ import { OpportunityTabExtension } from '../manifest/extensions/tabs/types/Oppor
...
 import { StoreType } from '../manifest/store/StoreType';
 import { Locale } from '../sdk/Locale';
 import { ManifestV1 } from './ManifestV1';
 
 export class ManifestTranslator {
-  public static getAddonManifest(
-    app: Application,
-    ext: TabExtension | ShellExtension
-  ): ManifestV1 | null {
+  public static getAddonManifest(app: Application, ext: TabExtension | ShellExtension): ManifestV1 | null {
```
Comment (palootcenas-outreach) [inline src/legacy/ManifestTranslator.ts:43]: It allows providing only `ext: TabExtension | ShellExtension` but in code below we're checking a lot more allowed `ext.type`s

File: src/sdk/services/oauthService.ts (hunk 8 above)
Comment (palootcenas-outreach) [reply src/sdk/services/oauthService.ts]: This is related to [the comment](https://github.com/getoutreach/extensibility-sdk/pull/31#discussion_r878063506) that would make all deprecated props optional

File: src/sdk/services/oauthService.ts (hunk 6 above)
Comment (palootcenas-outreach) [reply src/sdk/services/oauthService.ts]: No it was not, space and plus are the same character in URL.

Comment (sheff146) [reply src/sdk/services/oauthService.ts]: Wow, I didn't know that. Let me revert it to how it was then.

File: src/legacy/ManifestTranslator.ts (hunk 11 above)
Comment (sheff146) [reply src/legacy/ManifestTranslator.ts:43]: I see only `ShellExtensionType` and `TabExtensionType` checks.

File: src/sdk/services/oauthService.ts (hunk 6 above)
Comment (palootcenas-outreach) [reply src/sdk/services/oauthService.ts]: Actually [it's more complicated](https://stackoverflow.com/a/47188851). 
TL;DR: The plus should work, but using space was actually more correct according to [RFC3986](https://www.rfc-editor.org/rfc/rfc3986)

File: src/legacy/ManifestTranslator.ts (hunk 11 above)
Comment (palootcenas-outreach) [reply src/legacy/ManifestTranslator.ts:43]: You are right.

File: src/manifest/ManifestApi.ts (hunk 9 above)
Comment (sheff146) [reply src/manifest/ManifestApi.ts:51]: I'm trying it right now and it brings additional complications to validation. Also, some of those fields are used in SDK methods. I'd rather leave them as they are now and remove them completely later. On our side we can take care about filling all of them with values.

File: src/sdk/Validator.ts (hunk 5 above)
Comment (sheff146) [reply src/sdk/Validator.ts]: Added a condition

File: src/manifest/ManifestApi.ts (hunk 9 above)
Comment (palootcenas-outreach) [reply src/manifest/ManifestApi.ts:70]: Ok

File: src/sdk/services/oauthService.ts (hunk 10 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts:4]: Added couple of tests

File: docs/schema/2.0/manifest.schema.json (hunk 3 above)
Comment (sheff146) [reply docs/schema/2.0/manifest.schema.json:70]: TBH I seriously doubt that we use it somewhere and that it is up-to-date.

File: src/index.ts [hunk 12]
```diff
@@ -451,7 +451,7 @@ class ExtensibilitySdk {
    * @returns {Promise<string | null>}
    * @memberof ExtensibilitySdk
    */
-  public authenticate = async (): Promise<string | null> => {
+  public authenticate = async (redirectUri?: string): Promise<string | null> => {
```
Comment (antonaudzei-outreach) [inline src/index.ts:454]: There is one more optional parameter to append to popup URL that I believe will be useful especially in case of multiple redirectUris: [state parameter](https://medium.com/keycloak/the-importance-of-the-state-parameter-in-oauth-5419c94bef4c)

I therefore suggest adding it as optional here as well

File: src/sdk/services/oauthService.ts [hunk 13]
```diff
@@ -2,33 +2,48 @@ import { ManifestApi } from '../../manifest/ManifestApi';
...
+    return `${host}/oauth/authorize?client_id=${clientId}&redirect_uri=${selectedRedirectUri}&response_type=code&scope=${scopes}`;
+  };
+
+  private selectRedirectUri = (api: ManifestApi, redirectUri?: string): string => {
+    if (redirectUri === undefined) {
+      if (api.redirectUris?.length) {
+        return api.redirectUris[0];
+      }
+
+      return api.redirectUri ?? '';
```
Comment (nimal-outreach) [inline src/sdk/services/oauthService.ts:39]: If api.redirectUri is undefined here you will return empty string - will this be valid value? in L46 you throw and not return empty string?

File: src/manifest/ManifestApi.ts [hunk 14]
```diff
@@ -9,11 +10,42 @@ import { Scopes } from './api/Scopes';
...
+   * @memberof ManifestApi
+   */
+  redirectUris: string[];
+
   /**
    * Outreach OAuth application id
    *
    * @see https://github.com/getoutreach/extensibility-sdk/blob/master/docs/manifest.md#applicationid
    * @type {string}
+   * @deprecated applicationId is deprecated and replaced by client.id
```
Comment (nimal-outreach) [inline src/manifest/ManifestApi.ts:48]: Why not just removing this? 

The type of change this PR brings is of so significant  that we might draw a line in the sand here and say use client.* instead?

File: docs/manifest.md [hunk 15]
```diff
@@ -99,9 +101,10 @@ opportunity tab extension.
   },
   "api": {
     "scopes": ["accounts.all", "calls.all"],
-    "applicationId": "AbCd123456qW",
-    "redirectUri": "https://application-host.com/hello-world",
-    "connect": "https://someurl.com/connect"
+    "client": {
```
Comment (nimal-outreach) [inline docs/manifest.md:104]: What is the reason we have here client: { id } vs having it just as clientId (renaming existing applicationId)?

**Review by nimal-outreach — APPROVED** [review]
Approved with few comments/suggestions - please check them out

File: src/manifest/ManifestApi.ts (hunk 14 above)
Comment (sheff146) [reply src/manifest/ManifestApi.ts:48]: If you say we can safely delete deprecated fields from the api type, I will gladly do so.

File: docs/manifest.md (hunk 15 above)
Comment (sheff146) [reply docs/manifest.md:104]: Just emphasizing that this is a separate entity and leaving a space for potential additions in a future. Would you prefer to have just `clientId` field?

File: src/sdk/services/oauthService.ts (hunk 13 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts:43]: In L46 a parameter provided by a developer doesn't match any of redirectUris, that's why we throw. Here it's different.

File: src/manifest/ManifestApi.ts (hunk 14 above)
Comment (palootcenas-outreach) [reply src/manifest/ManifestApi.ts:48]: I think this would be a breaking change that would require bumping up the manifest version as well as the SDK package major version. 
But yes, we can do it 😉

File: src/sdk/services/oauthService.ts (hunk 6 above)
Comment (palootcenas-outreach) [reply src/sdk/services/oauthService.ts]: I can still see comma in code, but the comment is resolved. You said you're going to revert it, i.e. space will be used.

File: src/sdk/services/oauthService.ts (hunk 7 above)
Comment (palootcenas-outreach) [reply src/sdk/services/oauthService.ts]: I don't see the fix yet

File: src/manifest/ManifestApi.ts (hunk 14 above)
Comment (sheff146) [reply src/manifest/ManifestApi.ts:48]: Rather not in this PR

File: src/sdk/services/oauthService.ts (hunk 7 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts]: You won't see it in the conversation, you have to go though the PR files again. Github UX is not ideal 👎

File: src/sdk/services/oauthService.ts (hunk 6 above)
Comment (sheff146) [reply src/sdk/services/oauthService.ts]: I have reverted it.
